### ISSUE TITLE
Chef doesn't fail if a resource's required properties is missing

### ISSUE THREAD

**Issue by failshell** [chef/chef#4596]
## Description

I'm writing a new resource in which i defined required properties. When using the resource and omitting the required attributes, Chef happily continues its run without failing. I would expect it to fail.
## Chef Version

I tested with 12.5.1 and 12.7.2.
## Platform Version

Centos 7.1
## Replication Case

Here's the resource file:

``` text
resource_name :rundeck_user

property :password, String, required: true
property :roles, Array, required: true

action :create do
  puts 'foo'
end
```

And how it's called inside another cookbook:

``` text
rundeck_user 'test'
```
## Stacktrace

As seen https://gist.github.com/failshell/17553f8b90287f5a6da8, if i raise then I get the expected result.

**Comment by coderanger** [comment]
This has always been slightly confusing, but the behavior hasn't changed. The `required` flag is only checked on the first time the property is used. If you change the action code to `puts new_resource.roles` it would fail as expected.

@lamont-granquist This is definitely not an `accepted minor` IMO, if we wanted to change this behavior it would have to be in a major release.

**Comment by lamont-granquist** [comment]
Wow, that's a bit lolwut.

Lets Accepted Major this and drop the priority down to Low.

**Comment by failshell** [comment]
@coderanger ah cool, good to know. thought i was going nuts here :)

**Comment by coderanger** [comment]
:+1: It's definitely something that catches a lot of people off guard.

**Comment by failshell** [comment]
Strangely, I've been writing resources for years and never thought to check if that behavior worked before accessing the actually attribute.

Live and learn :)

**Comment by sekrett** [comment]
I think old behavior is good. For example I have many properties in my resource, all of them are required for a `create` action, and only one for `destroy` action. If a property is not used, why Chef should fail? Any opinions?

**Comment by lamont-granquist** [comment]
yeah this might be a feature.  the validation has to be lazy in order to deal with explicitly lazy properties and `coerce:` blocks and stuff.

the end result is that if you never use a property then it isn't validated.  the workaround is just to reference the property (in a void context should work) which will cause it to get accessed and validated.

i definitely don't see a way forwards to fixing this in Chef-13

**Comment by lamont-granquist** [comment]
need to evaluate this one again in time for Chef-16 to see if there's any solution

**Comment by lamont-granquist** [comment]
So probably the fix to this should be to before every action taken to run through all the properties which are required and poke them and cause them to fail validation if they're not set.

But that means that we probably need to support more than boolean true/false but also a list of symbols corresponding to actions where the property is required:

```
property :whatever, String, required: %i{enable disable}  # not required on :start and :stop actions
```

**Comment by lamont-granquist** [comment]
Fixed in #9688 which is a breaking change.

The people that have been relying on this behavior where required properties were not required if they weren't referenced, will need to push an array of action symbols into the required option to fix that.
